rfcs/20201201-cpp-gradients.md

```diff
@@ -0,0 +1,415 @@
+    exp_output->Unref();
+  }
+  return Status::OK();
+}
+```
+
+
+
+#### tf.recompute\_grad
+
+`tf.recompute_grad` is an application of `tf.custom_gradient` where we do not record the forward pass on the tape so that we are not holding on to forward pass tensors in memory. (In `tf.custom_gradient` we allow recording the forward pass on the tape in order for higher-order derivatives to work for cases where the custom gradient function uses intermediate tensors from the forward pass.) This is implemented by executing the forward pass outside the tape (managed by a higher layer) and registering a gradient function that re-runs the forward pass and computes gradients. The same behavior can be achieved using this tape.
+
+
+#### Nested tapes and higher-order derivatives
+
+Higher order derivatives are computed by either using a persistent tape or by tracing computations on nested tapes. The nesting is managed by a higher layer. This can be implemented using this tape.
+
+
+#### Skipping gradients for certain op inputs (skip\_input\_indices)
+
+A [small set](https://cs.opensource.google/search?q=f:py$%20skip_input_indices&sq=&ss=tensorflow%2Ftensorflow) of python gradient functions have been optimized to not return gradients for inputs which are not tracked under the tape. This is beneficial in eager mode where unneeded gradients cannot be pruned during execution. In the C++ tape, we support this by providing a `skip_input_indices` field on the `ForwardOperation` which stores the list of input indices which are not watched. 
+
+
+#### Automatic variable tracking
+
+In python, if a variable is accessed inside a `tf.GradientTape`s scope it is automatically tracked, i.e. `Tape::Watch` is called for the `DT_RESOURCE` tensor backing the variable on behalf of the user. For now we will leave this out as a higher layer feature and require that variable handles are explicitly tracked by a higher layer. We can revisit this later if needed.
+
+
+#### tf.function and functional control flow gradients [out of scope for now]
+
+Eventually we plan to implement tf.function and functional control flow gradients in C++ but for now we will leave them in python and directly call the python gradient function.
+
+
+#### IndexedSlices
+
+Gradient function of a [gather](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/ops/array_grad.py;l=582;drc=d724cdbce69862cbb80617dd6573baa83bd3e819) returns `IndexedSlices` for efficiency. We need to support `IndexedSlices` as part of the input and output gradients of a gradient function. Currently there is no good C++ representation for these. One possible representation would be to wrap the component tensors in an `IndexedSlicesTensorHandle` that subclasses `AbstractTensorHandle`. This way IndexedSlices would be transparent to the tape. The C++ gen ops can choose to handle `IndexedSlices` appropriately or simply densify them by calling a C++ equivalent of `convert_to_tensor`.
```

Comment:
Is this related to the FullType stuff I saw floating around elsewhere? (If we had a C++ representation for CompositeTensors/Extension types we could possibly just return those)

Reply:
Yes, I ran this idea by @mdanatg. We will possibly want to implement something like `CompositeTensorHandle` and subclass `IndexedSlicesTensorHandle` from that. The main takeaway here is that the tape API can support non-dense tensors by representing those as subclasses of `AbstractTensorHandle`.